File: android/src/main/java/cn/hecom/avatar/PathUtil.java
package cn.hecom.avatar;

import android.graphics.Path;
import android.graphics.PointF;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by kevin.bai on 2019-08-28. 计算过程参考{@see <a href="https://www.jianshu.com/p/edbe67f14751">博客</a>}
 */
public class PathUtil {
    public static void calculatePath(Path path, PointF center, int side, float size, float corner, float offset) {
        if (corner > 0) {
            calculatePathWithCorner(path, center, side, size, corner, offset);
        } else {
            calculatePath(path, center, side, size, offset);
        }
    }

    private static void calculatePath(Path path, PointF center, int side, float size, float offset) {
        List<PointF> points = regularPolygonCoordinates(center, side, size, offset);
        path.reset();
        for (int i = 0; i < points.size(); i++) {
            PointF point = points.get(i);
            if (i == 0) {
                path.moveTo(point.x, point.y);
            } else {
                path.lineTo(point.x, point.y);
            }
        }
        path.close();
    }

    private static void calculatePathWithCorner(Path path, PointF center, int side, float size, float corner,
                                                float offset) {
        List<PointF> points = regularPolygonCoordinatesWithRoundedCorner(center, side, size, corner, offset);
        PointF temPoint = null;
        for (int i = 0; i < points.size(); i++) {
            PointF p = points.get(i);
            if (i == 0) {
                path.moveTo(p.x, p.y);
            } else {
                switch (i % 3) {
                    case 0:
                        path.lineTo(p.x, p.y);
                        break;
                    case 1:
                        temPoint = p;
                        break;
                    case 2:
                        path.quadTo(temPoint.x, temPoint.y, p.x, p.y);
                        break;
                }
            }
        }
        path.close();
    }

    private static List<PointF> regularPolygonCoordinatesWithRoundedCorner(PointF center, int side, float radius,
                                                                           float corner, float offset) {
        List<PointF> points = new ArrayList<>();
        double CAB = 2 * Math.PI / side;
        double EC = radius * Math.sin((CAB / 2));
        double AE = radius * Math.cos((CAB / 2));
        double ED = EC - corner;
        double EAD = Math.atan(ED / AE);
        float DAC = (float) (CAB / 2 - EAD);
        float AD = (float) Math.sqrt(Math.pow(AE, 2) + Math.pow(ED, 2));
        for (int i = 0; i < side; i++) {
            float angle = (float) (i * CAB);
            PointF point = vertexCoordinates(center, radius, angle, offset);
            PointF leftPoint = vertexCoordinates(center, AD, angle - DAC, offset);
            PointF rightPoint = vertexCoordinates(center, AD, angle + DAC, offset);
            points.add(leftPoint);
            points.add(point);
            points.add(rightPoint);
        }
        return points;
    }

    private static List<PointF> regularPolygonCoordinates(PointF center, int side, float radius, float offset) {
        List<PointF> points = new ArrayList<>();
        for (int i = 0; i < side; i++) {
            float angle = (float) (i * (360.0f / side) / 180 * Math.PI);
            PointF point = vertexCoordinates(center, radius, angle, offset);
            points.add(point);
        }
        return points;
    }

    private static PointF vertexCoordinates(PointF center, float radius, float angle, float offset) {
        return new PointF(center.x + radius * (float) Math.cos(angle + offset), center.y + radius * (float) Math
                .sin(angle + offset));
    }
}

File: PathUtil.ts
interface PointF {
    x: number;
    y: number;
}

export function calculatePath(
    center: PointF,
    side: number,
    size: number,
    corner: number,
    offset: number
) {
    if (corner > 0) {
        return calculatePathWithCorner(center, side, size, corner, offset);
    } else {
        return calculatePathWithoutCorner(center, side, size, offset);
    }
}

function calculatePathWithoutCorner(
    center: PointF,
    side: number,
    size: number,
    offset: number
): string {
    const points: Array<PointF> = regularPolygonCoordinates(center, side, size, offset);
    let path: string = '';
    for (let i = 0; i < points.length; i++) {
        const p: PointF = points[i];
        if (i == 0) {
            // path.moveTo(point.x, point.y);
            path = `${path} M ${p.x}, ${p.y} `;
        } else {
            // path.lineTo(point.x, point.y);
            path = `${path} L ${p.x}, ${p.y} `;
        }
    }
    // path.close();
    path = `${path} Z `;

    return path;
}

function calculatePathWithCorner(
    center: PointF,
    side: number,
    size: number,
    corner: number,
    offset: number
): string {
    const points: Array<PointF> = regularPolygonCoordinatesWithRoundedCorner(
        center,
        side,
        size,
        corner,
        offset
    );
    let path: string = '';
    let temPoint: PointF = null;
    for (let i = 0; i < points.length; i++) {
        const p: PointF = points[i];
        if (i == 0) {
            // path.moveTo(p.x, p.y);
            path = `${path} M ${p.x}, ${p.y} `;
        } else {
            switch (i % 3) {
                case 0:
                    // path.lineTo(p.x, p.y);
                    path = `${path} L ${p.x}, ${p.y} `;
                    break;
                case 1:
                    temPoint = p;
                    break;
                case 2:
                    // path.quadTo(temPoint.x, temPoint.y, p.x, p.y);
                    path = `${path} Q ${temPoint.x}, ${temPoint.y}, ${p.x}, ${p.y} `;
                    break;
                default:
                    break;
            }
        }
    }
    // path.close();
    path = `${path} Z `;
    return path;
}

export function regularPolygonCoordinatesWithRoundedCorner(
    center: PointF,
    side: number,
    radius: number,
    corner: number,
    offset: number
): Array<PointF> {
    const points: Array<PointF> = [];
    const CAB = (2 * Math.PI) / side;
    const EC = radius * Math.sin(CAB / 2);
    const AE = radius * Math.cos(CAB / 2);
    const ED = EC - corner;
    const EAD = Math.atan(ED / AE);
    const DAC = CAB / 2 - EAD;
    const AD = Math.sqrt(AE ** 2 + ED ** 2);
    for (let i = 0; i < side; i++) {
        const angle = i * CAB;
        const point = vertexCoordinates(center, radius, angle, offset);
        const leftPoint = vertexCoordinates(center, AD, angle - DAC, offset);
        const rightPoint = vertexCoordinates(center, AD, angle + DAC, offset);
        points.push(leftPoint);
        points.push(point);
        points.push(rightPoint);
    }
    return points;
}

export function regularPolygonCoordinates(
    center: PointF,
    side: number,
    radius: number,
    offset: number
): Array<PointF> {
    const points: Array<PointF> = [];
    for (let i = 0; i < side; i++) {
        const angle = ((i * (360.0 / side)) / 180) * Math.PI;
        const point: PointF = vertexCoordinates(center, radius, angle, offset);
        points.push(point);
    }
    return points;
}

function vertexCoordinates(center: PointF, radius: number, angle: number, offset: number): PointF {
    return {
        x: center.x + radius * Math.cos(angle + offset),
        y: center.y + radius * Math.sin(angle + offset),
    };
}
